runners/core-java/src/test/java/org/apache/beam/runners/core/UnboundedReadFromBoundedSourceTest.java

```diff
@@ -0,0 +1,381 @@
+    testBoundedToUnboundedSourceAdapterCheckpointRestart(source, expected);
+  }
+
+  private <T> void testBoundedToUnboundedSourceAdapterCheckpointRestart(
+      BoundedSource<T> boundedSource,
+      List<T> expectedElements) throws Exception {
+    BoundedToUnboundedSourceAdapter<T> unboundedSource =
+        new BoundedToUnboundedSourceAdapter<>(boundedSource);
+
+    PipelineOptions options = PipelineOptionsFactory.create();
+    BoundedToUnboundedSourceAdapter<T>.Reader reader =
+        unboundedSource.createReader(options, null);
+
+    List<T> actual = Lists.newArrayList();
+    for (boolean hasNext = reader.start(); hasNext;) {
+      actual.add(reader.getCurrent());
+      // checkpoint every 9 elements
+      if (actual.size() % 9 == 0) {
+        Checkpoint<T> checkpoint = reader.getCheckpointMark();
+        Coder<Checkpoint<T>> checkpointCoder = unboundedSource.getCheckpointMarkCoder();
+        Checkpoint<T> decodedCheckpoint = CoderUtils.decodeFromByteArray(
+            checkpointCoder,
+            CoderUtils.encodeToByteArray(checkpointCoder, checkpoint));
+        reader.close();
+        checkpoint.finalizeCheckpoint();
+
+        BoundedToUnboundedSourceAdapter<T>.Reader restarted =
+            unboundedSource.createReader(options, decodedCheckpoint);
+        reader = restarted;
+        hasNext = reader.start();
+      } else {
+        hasNext = reader.advance();
+      }
+    }
+    assertEquals(expectedElements.size(), actual.size());
+    assertEquals(Sets.newHashSet(expectedElements), Sets.newHashSet(actual));
+  }
+
+  @Test
+  public void testReadBeforeStart() throws Exception {
+    thrown.expect(NoSuchElementException.class);
+
+    BoundedSource<Long> countingSource = CountingSource.upTo(100);
+    BoundedToUnboundedSourceAdapter<Long> unboundedSource =
+        new BoundedToUnboundedSourceAdapter<>(countingSource);
+    PipelineOptions options = PipelineOptionsFactory.create();
+
+    unboundedSource.createReader(options, null).getCurrent();
+  }
+
+  @Test
+  public void testReadFromCheckpointBeforeStart() throws Exception {
+    thrown.expect(NoSuchElementException.class);
+
+    BoundedSource<Long> countingSource = CountingSource.upTo(100);
+    BoundedToUnboundedSourceAdapter<Long> unboundedSource =
+        new BoundedToUnboundedSourceAdapter<>(countingSource);
+    PipelineOptions options = PipelineOptionsFactory.create();
+
+    List<TimestampedValue<Long>> elements =
+        ImmutableList.of(TimestampedValue.of(1L, new Instant(1L)));
+    Checkpoint<Long> checkpoint = new Checkpoint<>(elements, countingSource);
+    unboundedSource.createReader(options, checkpoint).getCurrent();
+  }
+
+  /**
+   * Generate byte array of given size.
+   */
+  private static byte[] generateInput(int size) {
+    // Arbitrary but fixed seed
+    Random random = new Random(285930);
+    byte[] buff = new byte[size];
+    for (int i = 0; i < size; i++) {
+      buff[i] = (byte) (random.nextInt() % Byte.MAX_VALUE);
```

Comment:
Use the change here -- and see comments for why. https://github.com/apache/incubator-beam/pull/456